File: haskell/basics/fibonacci.hs
-- Print the first 20 Fibonacci numbers from an infinite self-referential list.
-- Run: runghc fibonacci.hs

fibonacci :: [Integer]
fibonacci = 0 : 1 : zipWith (+) fibonacci (tail fibonacci)

main :: IO ()
main = putStrLn . unwords . map show $ take 20 fibonacci

File: kotlin/basics/Fibonacci.kt
// Print the first 20 Fibonacci numbers from a lazily generated sequence.
// Run: kotlinc Fibonacci.kt -include-runtime -d fibonacci.jar && java -jar fibonacci.jar

fun fibonacci(): Sequence<Long> =
    generateSequence(0L to 1L) { (current, next) -> next to (current + next) }
        .map { (current, _) -> current }

fun main() {
    println(fibonacci().take(20).joinToString(" "))
}
